cirq-core/cirq/experiments/readout_confusion_matrix.py

```diff
@@ -0,0 +1,310 @@
+
+        The single `2 ** len(qubits) x 2 ** len(qubits)` confusion matrix is constructed
+        using the individual smaller `self.confusion_matrices` by applying necessary
+        matrix transpose / kron / partial trace operations.
+
+        Args:
+            qubits: The qubits representing the subspace for which a confusion matrix should be
+                    constructed. By default, uses all qubits in sorted order, i.e. `self.qubits`.
+
+        Returns:
+            Confusion matrix for subspace corresponding to `qubits`.
+
+        Raises:
+            ValueError: If `qubits` is not a subset of `self.qubits`.
+        """
+
+        if qubits is None:
+            qubits = self.qubits
+        if any(q not in self.qubits for q in qubits):
+            raise ValueError(f"qubits {qubits} should be a subset of self.qubits {self.qubits}.")
+        return self._confusion_matrix(tuple(qubits))
+
+    def correction_matrix(self, qubits: Optional[Sequence['cirq.Qid']] = None) -> np.ndarray:
+        """Returns a single correction matrix constructed for the given set of qubits.
+
+        A correction matrix is the inverse of confusion matrix and can be used to apply corrections
+        to observed frequencies / probabilities to compensate for the readout error.
+        A Moore–Penrose Pseudo inverse of the confusion matrix is computed to get the correction
+        matrix.
+
+        Args:
+            qubits: The qubits representing the subspace for which a correction matrix should be
+                    constructed. By default, uses all qubits in sorted order, i.e. `self.qubits`.
+
+        Returns:
+            Correction matrix for subspace corresponding to `qubits`.
+
+        Raises:
+            ValueError: If `qubits` is not a subset of `self.qubits`.
+        """
+
+        if qubits is None:
+            qubits = self.qubits
+        if any(q not in self.qubits for q in qubits):
+            raise ValueError(f"qubits {qubits} should be a subset of self.qubits {self.qubits}.")
+        return np.linalg.pinv(self.confusion_matrix(qubits))
+
+    def apply(
+        self,
+        result: np.ndarray,
+        qubits: Optional[Sequence['cirq.Qid']] = None,
+        *,
+        method='least_squares',
+    ) -> np.ndarray:
+        """Applies corrections to the observed `result` to compensate for readout error on qubits.
+
+        The compensation is applied by multiplying the result with the correction matrix
+        corresponding to the subspace defined by `qubits`.
+
+        Args:
+            result: `(2 ** len(qubits), )` shaped numpy array containing observed frequencies /
+                    probabilities.
+            qubits: Sequence of qubits used for sampling to get `result`. By default, uses all
+                    qubits in sorted order, i.e. `self.qubits`.
+            method: Correction Method. Should be either 'pseudo_inverse' or 'least_squares'.
```

Comment:
constrained least squares. describe constraint

Reply:
added explanation on top. Sticking to 'least_squares' instead of 'constrained_least_squares' to keep the name short.

Reply:
thanks